### DISCUSSION TITLE
How many branching decisions separate the root node from the first node enumerated?

### DISCUSSION THREAD

**Discussion by spkelle2** [coin-or/Cbc#552] [Q&A]
I ran p0201.mps with preprocess and presolve off and CBC enumerated 59 nodes. Upon inspecting each node, I find that the 1st and 31st nodes to be processed are roots of their own branch and bound (sub?)trees (i.e. all nodes have exactly one of those two nodes as an ancestor), but none of the enumerated nodes is a common ancestor to all other nodes. Further inspection of their CbcBranchingObject.variable_ and CbcBranchingObjects.way_ shows the 1st and 31st nodes to be created from branching on the same variable (x_10) but in different directions. Both the 1st and 31st nodes have several variables with tightened bounds relative to the LP relaxation for p0201.mps. Can I assume that x_10 is the only variable in the 1st and 31st nodes with bounds tightened due to branching? With only preprocess and presolve off, can I assume that all other tightened variable bounds are due to feasibility cuts found for p0201.mps? If not, where else in CBC causes the tightening of these variables' bounds?

The reason I ask the first question above is for the following. Next, I ran 10teams.mps from MIPLIB3 with preprocess and presolve off and CBC enumerated 32 nodes. Upon inspecting each node, I find that all nodes have the 1st processed node as an ancestor, which makes me think it was the root node of the branch and bound tree. However, when I check CbcBranchingObject.variable_ for the first node, I see it was created from branching on x_1618. Why then isn't there a disjoint (sub)tree for the node that was created by branching in the opposite direction?

**Comment by jjhforrest** [comment]
About p0201 - when I run I get a line at end
Cbc0032I Strong branching done 584 times (11926 iterations), fathomed 8 nodes and fixed 7 variables
so strong branching can also tighten bounds.
As to 10teams, the answer is probably that cuts at root node changed objective to 924.0.  A solution of 924 was found, so the branch and bound algorithm terminates - there is no point branching other way on root node.

**Reply by spkelle2** [reply]
Ok. I think I'm clear on where the root node is after your response and revisiting my command line output. Thank you for the follow up.

I think that leave just one question I have from above that is outstanding. Can you clarify how I can tell which of the remaining tightened variable bounds are from feasibility cuts and which are from optimality cuts?
